service/history/historyReplicator.go

```diff
@@ -68,167 +77,238 @@ func newHistoryReplicator(shard ShardContext, historyEngine *historyEngineImpl,
-				rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID)
-			previousActiveCluster := r.metadataMgr.ClusterNameForFailoverVersion(rState.LastWriteVersion)
-			ri, ok := request.ReplicationInfo[previousActiveCluster]
-			if !ok {
-				logger.Errorf("No replication information found for previous active cluster.  Previous: %v, Request: %v, ReplicationInfo: %v",
-					previousActiveCluster, request.GetSourceCluster(), request.ReplicationInfo)
-
-				// TODO: Handle missing replication information.
-				// Returning BadRequestError to force the message to land into DLQ
-				return ErrMissingReplicationInfo
-			}
+func (r *historyReplicator) ApplyStartEvent(context *workflowExecutionContext, request *h.ReplicateEventsRequest, logger bark.Logger) error {
+	msBuilder := r.getNewMutableState(request.GetVersion(), logger)
+	err := r.ApplyReplicationTask(context, msBuilder, request, logger)
+	if err != nil {
+		logger.Errorf("Fail to Apply Replication task.  NextEvent: %v, FirstEvent: %v, Err: %v", msBuilder.GetNextEventID(),
+			request.GetFirstEventId(), err)
+	}
+	return err
+}
 
-			// Detect conflict
-			if ri.GetLastEventId() != rState.LastWriteEventID {
-				r.metricsClient.IncCounter(metrics.ReplicateHistoryEventsScope, metrics.HistoryConflictsCounter)
-				logger.Infof("Conflict detected.  State: {V: %v, LastWriteV: %v, LastWriteEvent: %v}, ReplicationInfo: {PrevC: %v, V: %v, LastEvent: %v}",
-					rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID,
-					previousActiveCluster, ri.GetVersion(), ri.GetLastEventId())
-
-				resolver := newConflictResolver(r.shard, context, r.historyMgr, r.logger)
-				msBuilder, err = resolver.reset(uuid.New(), ri.GetLastEventId(), msBuilder.GetExecutionInfo().StartTimestamp)
-				logger.Infof("Completed Resetting of workflow execution.  NextEventID: %v. Err: %v", msBuilder.GetNextEventID(), err)
-				if err != nil {
-					return err
-				}
-			}
-		}
+func (r *historyReplicator) ApplyOtherEventsMissingMutableState(domainID string, workflowID string, incomingVersion int64, logger bark.Logger) error {
+	// we need to check the current workflow execution
+	currentContext, release, err := r.historyCache.getOrCreateWorkflowExecution(
+		domainID,
+		// only use the workflow ID, to get the current running one
+		shared.WorkflowExecution{WorkflowId: common.StringPtr(workflowID)},
+	)
+	if err != nil {
+		return err
+	}
+	currentMsBuilder, err := currentContext.loadWorkflowExecution()
+	if err != nil {
+		// no matter what error happen, we need to retry
+		release(err)
+		return err
+	}
+	currentLastWriteVersion := currentMsBuilder.GetLastWriteVersion()
+	currentRunID := currentMsBuilder.GetExecutionInfo().RunID
+	release(nil)
+
+	// we can also use the start version
+	if currentLastWriteVersion > incomingVersion {
+		logger.Infof("Dropping replication task. Current RunID: %v, Current LastWriteVersion: %v, Incoming Version: %v.",
```

Comment:
increment the dropped counter.

Reply:
added all 3 types:
HistoryConflictsCounter
StaleReplicationEventsCounter
BufferedReplicationTaskCounter